src/boot_O2/__osMalloc.c

```diff
@@ -179,45 +186,91 @@ void __osFree(Arena* arena, void* ptr) {
     ArenaNode* node;
     ArenaNode* next;
     ArenaNode* prev;
-    ArenaNode* newNext;
 
     ArenaImpl_Lock(arena);
-    node = (ArenaNode*)((u32)ptr - sizeof(ArenaNode));
 
-    if (ptr == NULL || (node->magic != NODE_MAGIC) || node->isFree) {
-        goto end;
-    }
+    node = (ArenaNode*)((uintptr_t)ptr - sizeof(ArenaNode));
+
+    if ((ptr != NULL && (node->magic == NODE_MAGIC) && !node->isFree)) {
+        next = node->next;
+        prev = node->prev;
+        node->isFree = true;
+
+        if ((uintptr_t)next == (uintptr_t)node + sizeof(ArenaNode) + node->size && next->isFree) {
+            ArenaNode* newNext = next->next;
+
+            if (newNext != NULL) {
+                newNext->prev = node;
+            }
 
-    next = node->next;
-    prev = node->prev;
-    node->isFree = 1;
+            node->size += next->size + sizeof(ArenaNode);
 
-    newNext = next;
-    if ((u32)next == (u32)node + sizeof(ArenaNode) + node->size && next->isFree) {
-        newNext = next->next;
-        if (newNext != NULL) {
-            newNext->prev = node;
+            node->next = newNext;
+            next = newNext;
         }
 
-        node->size += next->size + sizeof(ArenaNode);
+        if (prev != NULL && prev->isFree && (uintptr_t)node == (uintptr_t)prev + sizeof(ArenaNode) + prev->size) {
+            if (next != NULL) {
+                next->prev = prev;
+            }
 
-        node->next = newNext;
-        next = newNext;
+            prev->next = next;
+            prev->size += node->size + sizeof(ArenaNode);
+        }
     }
 
-    if (prev != NULL && prev->isFree && (u32)node == (u32)prev + sizeof(ArenaNode) + prev->size) {
-        if (next) {
-            next->prev = prev;
+    ArenaImpl_Unlock(arena);
+}
+
+void* __osRealloc(Arena* arena, void* ptr, size_t newSize) {
+    ArenaImpl_Lock(arena);
+
+    (void)"__osRealloc(%08x, %d)\n";
+
+    if (ptr == NULL) {
+        ptr = __osMalloc(arena, newSize);
+    } else if (newSize == 0) {
+        __osFree(arena, ptr);
+        ptr = NULL;
+    } else {
+        u32 diff;
+        void* newPtr;
+        ArenaNode* node;
+
+        node = (uintptr_t)ptr - sizeof(ArenaNode);
+        newSize = ALIGN16(newSize);
+        if ((newSize != node->size) && (node->size < newSize)) {
+            ArenaNode* next = node->next;
+
+            diff = newSize - node->size;
+            if (((uintptr_t)next == ((uintptr_t)node + node->size + sizeof(ArenaNode))) && (next->isFree) &&
```

Comment:
Since this is the fancy bit of realloc, could you annotate what this is doing a bit?